src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,316 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/support/PersistentData.h>
+
+namespace chip {
+
+enum class PersistentTags : uint8_t
+{
+    kCount   = 1,
+    kList    = 2,
+    kEntries = 3,
+};
+
+template <size_t kMaxSerializedSize, size_t kMaxListSize, typename EntryType>
+struct PersistentArray : public PersistentData<kMaxSerializedSize>
+{
+
+    PersistentArray(PersistentStorageDelegate * storage) : PersistentData<kMaxSerializedSize>(storage), mLimit(kMaxListSize) {}
+
+    size_t Limit() { return this->mLimit; }
+    uint16_t Count() { return this->mCount; }
+    virtual void Clear(EntryType & entry)                                                = 0;
+    virtual CHIP_ERROR Serialize(TLV::TLVWriter & writer, const EntryType & entry) const = 0;
+    virtual CHIP_ERROR Deserialize(TLV::TLVReader & reader, EntryType & entry)           = 0;
+
+    virtual CHIP_ERROR Copy(EntryType & target, const EntryType & source)
```

Comment:
Again, this needs a lot of documentation about what it's meant to do.  I don't think the way we use this below is safe.